File: sdks/node/src/webhook.ts
import { createHmac, timingSafeEqual } from 'crypto';

/**
 * Verify a webhook signature from Jobcelis.
 *
 * @param secret - The webhook secret (from webhook configuration)
 * @param body - The raw request body as a string
 * @param signature - The value of the X-Signature header (format: "sha256=<base64>")
 * @returns true if the signature is valid
 *
 * @example
 * ```typescript
 * import { verifyWebhookSignature } from '@jobcelis/sdk';
 *
 * app.post('/webhook', (req, res) => {
 *   const isValid = verifyWebhookSignature(
 *     'your_webhook_secret',
 *     req.rawBody,
 *     req.headers['x-signature'] as string
 *   );
 *
 *   if (!isValid) {
 *     return res.status(401).send('Invalid signature');
 *   }
 *
 *   // Process webhook...
 *   res.status(200).send('OK');
 * });
 * ```
 */
export function verifyWebhookSignature(
  secret: string,
  body: string,
  signature: string
): boolean {
  if (!secret || !body || !signature) return false;

  const prefix = 'sha256=';
  if (!signature.startsWith(prefix)) return false;

  const receivedSig = signature.slice(prefix.length);
  const expectedSig = createHmac('sha256', secret)
    .update(body)
    .digest('base64')
    .replace(/=+$/, ''); // no padding

  try {
    const a = Buffer.from(receivedSig, 'base64');
    const b = Buffer.from(expectedSig, 'base64');
    if (a.length !== b.length) return false;
    return timingSafeEqual(a, b);
  } catch {
    return false;
  }
}

File: sdks/rust/src/webhook.rs
use base64::Engine;
use base64::engine::general_purpose::STANDARD_NO_PAD;
use hmac::{Hmac, Mac};
use sha2::Sha256;
use subtle::ConstantTimeEq;

type HmacSha256 = Hmac<Sha256>;

/// Verify a webhook signature using HMAC-SHA256 (Base64, no padding).
///
/// # Arguments
///
/// * `secret` - The webhook signing secret.
/// * `body` - The raw request body.
/// * `signature` - The signature from the X-Signature header (format: "sha256=<base64>").
///
/// # Returns
///
/// `true` if the signature is valid.
pub fn verify_webhook_signature(secret: &str, body: &str, signature: &str) -> bool {
    let prefix = "sha256=";
    let received_sig = match signature.strip_prefix(prefix) {
        Some(s) => s,
        None => return false,
    };

    let mut mac = match HmacSha256::new_from_slice(secret.as_bytes()) {
        Ok(m) => m,
        Err(_) => return false,
    };
    mac.update(body.as_bytes());
    let result = mac.finalize();
    let expected = STANDARD_NO_PAD.encode(result.into_bytes());

    expected.as_bytes().ct_eq(received_sig.as_bytes()).into()
}
